sceptre_cdk_handler/cdk_builder.py

```diff
@@ -0,0 +1,214 @@
+from typing import Any, Optional, Dict, Type
+
+import aws_cdk
+from aws_cdk.cx_api import CloudAssembly
+from botocore.credentials import Credentials
+from cdk_bootstrapless_synthesizer import BootstraplessStackSynthesizer
+from sceptre import exceptions
+from sceptre.connection_manager import ConnectionManager
+from sceptre.exceptions import TemplateHandlerArgumentsInvalidError
+
+
+class SceptreCdkStack(aws_cdk.Stack):
+    def __init__(self, scope: aws_cdk.App, id: str, sceptre_user_data: Any, **kwargs):
+        super().__init__(scope, id, **kwargs)
+        self.sceptre_user_data = sceptre_user_data
+
+
+class CdkBuilder(ABC):
+    STACK_LOGICAL_ID = 'CDKStack'
+
+    @abstractmethod
+    def build_template(
+        self,
+        stack_class: Type[SceptreCdkStack],
+        cdk_context: Optional[dict],
+        sceptre_user_data: Any
+    ) -> dict: ...
+
+
+class BootstrappedCdkBuilder(CdkBuilder):
+
+    def __init__(
+        self,
+        logger: logging.Logger,
+        connection_manager: ConnectionManager,
+        *,
+        subprocess_run=subprocess.run,
+        app_class=aws_cdk.App,
+        environment_variables=os.environ
+    ):
+        self._logger = logger
+        self._connection_manager = connection_manager
+        self._subprocess_run = subprocess_run
+        self._app_class = app_class
+        self._environment_variables = environment_variables
+
+    def build_template(
+        self,
+        stack_class: Type[SceptreCdkStack],
+        cdk_context: Optional[dict],
+        sceptre_user_data: Any
+    ) -> dict:
+        assembly = self._synthesize(stack_class, cdk_context, sceptre_user_data)
+        self._publish(assembly)
+        template = self._get_template(assembly)
+        return template
+
+    def _synthesize(
+        self,
+        stack_class: Type[SceptreCdkStack],
+        cdk_context: Optional[dict],
+        sceptre_user_data: Any
+    ) -> CloudAssembly:
+        self._logger.debug(f'CDK synthesing CdkStack Class')
+        self._logger.debug(f'CDK Context: {cdk_context}')
+        app = self._app_class(context=cdk_context)
+        stack_class(app, self.STACK_LOGICAL_ID, sceptre_user_data)
+        return app.synth()
+
+    def _publish(self, cloud_assembly: CloudAssembly):
+        asset_artifacts = self._get_assets_manifest(cloud_assembly)
+        if self._only_asset_is_template(asset_artifacts):
+            # Sceptre already has a mechanism to upload the template if configured. We don't
+            # need to deploy assets if the only asset is the template
+            self._logger.debug("Only asset is template; Skipping asset upload.")
+            return
+
+        environment_variables = self._get_envs()
+        self._logger.info(f'Publishing CDK assets')
+        self._logger.debug(f'Assets manifest file: {asset_artifacts.file}')
+        self._run_command(
+            f'npx cdk-assets -v publish --path {asset_artifacts.file}',
+            env=environment_variables
```

Comment:
i'm not sure if this will output a useful message if cdk-assets is not installed.  would it make sense to handle that exception here and remind users to install cdk-assets?

Reply:
That's a good point. I just tried using `npx cdk-assets publish` when cdk-assets isn't installed and it actually hangs, seeking input:

![image](https://user-images.githubusercontent.com/77296393/208716518-3e34035a-3020-420e-8404-b0f4d0a4d167.png)

@X-Guardian originally had some protections against this and I foolishly removed them. I guess I should have retained those. I'll bring them back for this command.

Reply:
I've brought back these protections from the original code and put them in the "CommandChecker", which is injected into the template handler and checked during the validation step.